cartesi-rollups/tutorials/ether-wallet.md

```diff
@@ -0,0 +1,426 @@
+
+
+## Using the Ether wallet
+
+Now, let's create a simple application in the entrypoint, `src/index.ts` to test the wallet functionality.
+
+The [`EtherPortal`](../rollups-apis/json-rpc/portals/EtherPortal.md) contract allows anyone to perform transfers of Ether to a dApp. All deposits to a dApp is done via the `EtherPortal` contract.
+
+Run `cartesi address-book` to get the contract address of the `EtherPortal` contract. Save this as a const in the `index.ts` file.
+
+```typescript
+import createClient from "openapi-fetch";
+import { components, paths } from "./schema";
+import { Wallet } from "./wallet/wallet";
+import { stringToHex, getAddress, Address, hexToString, toHex } from "viem";
+
+type AdvanceRequestData = components["schemas"]["Advance"];
+type InspectRequestData = components["schemas"]["Inspect"];
+type RequestHandlerResult = components["schemas"]["Finish"]["status"];
+type RollupsRequest = components["schemas"]["RollupRequest"];
+export type Notice = components["schemas"]["Notice"];
+export type Payload = components["schemas"]["Payload"];
+export type Report = components["schemas"]["Report"];
+export type Voucher = components["schemas"]["Voucher"];
+
+type InspectRequestHandler = (data: InspectRequestData) => Promise<void>;
+type AdvanceRequestHandler = (
+  data: AdvanceRequestData
+) => Promise<RequestHandlerResult>;
+
+const wallet = new Wallet();
+
+const EtherPortal = `0xFfdbe43d4c855BF7e0f105c400A50857f53AB044`;
+
+const rollupServer = process.env.ROLLUP_HTTP_SERVER_URL;
+console.log("HTTP rollup_server url is " + rollupServer);
+
+const handleAdvance: AdvanceRequestHandler = async (data) => {
+  console.log("Received advance request data " + JSON.stringify(data));
+
+  const sender = data["metadata"]["msg_sender"];
+  const payload = data.payload;
+
+  if (sender.toLowerCase() === EtherPortal.toLowerCase()) {
+    // Handle deposit
+    const deposit = wallet.depositEther(payload);
+    await sendNotice(stringToHex(deposit));
+  } else {
+    // Handle transfer or withdrawal
+    try {
+      const { operation, from, to, amount } = JSON.parse(hexToString(payload));
+
+      if (operation === "transfer") {
+        const transfer = wallet.transferEther(
+          getAddress(from as Address),
+          getAddress(to as Address),
+          BigInt(amount)
+        );
+        console.log(transfer);
+        await sendNotice(stringToHex(transfer));
+      } else if (operation === "withdraw") {
+        const withdraw = wallet.withdrawEther(
+          getAddress(EtherPortal as Address),
+          getAddress(from as Address),
+          BigInt(amount)
+        );
+        console.log(withdraw);
+        await sendVoucher({ payload: amount, destination: from });
```

Comment:
on withdrawals
the `destination` you are using `from` value. It should not be the user id, but the dapp address relayed through the relay address call
```
await sendVoucher({ payload: amount, destination: from });
```

Reply:
Also the payload should not be an amount value, but the encoded call that you should return on the withdrawEther function